### ISSUE TITLE
Audio-echo and native-audio in the sample project do not work in Nexus 6P

### ISSUE THREAD

**Issue by lakdong** [googlesamples/android-ndk#367]
Audio-echo and native-audio in the sample project do not work in android nougat.

I tested it on the Nexus 6P ( Android Version 7.1.1 )

**Comment by ggfan** [comment]
thanks for reporting the issue, will take a look! Please list your build environment like buildToolsVersion etc in case some are automatically updated by your tool.

**Comment by lakdong** [comment]
My build environment
Android Studio 2.3
BuildToolsVersion 25.0.0
com.android.tools.build:Gradle:2.3.0
services.gradle.org/distributions/gradle-3.3-all.zip

**Comment by ggfan** [comment]
I saw it on Nexus 6, will investigate ( might be a little later: need to get other 2 things done ).

**Comment by Sathya-Subbiah** [comment]
Hi.. Audio-Echo sample is found not work in Android 7.1.1 version. Facing same issue. Any update on the above issue?

**Comment by ggfan** [comment]
Not yet, this one falling off, need to look at it later in the week.

**Comment by Sathya-Subbiah** [comment]
By increasing the buffersize the audio echo is heard in Nexus phone (Nexus 6 - Version 7.1.1). But heard with latency. Need some suggestion to reduce the latency.

**Comment by ggfan** [comment]
I need to have a look to see: the buffer size is in Fast-Audio path, once changed, it would be out of the fast audio path;  between input stream and output stream is a queue of buffers: could you increase the buffer count ( it is macro ) instead  -- if that works, the latency will increase, but it should still be within fast audio path.  Also there is one effect I need to disable to make input into the fast path ( just not having time to get to it ).  you might try them out ( you are in the center of "Pro Audio" war :-))

**Comment by IgorGanapolsky** [comment]
I am facing the same situation on Samsung Galaxy S4 running android 7.1.1.  When I press the start button, I see this stacktrace:
> libOpenSLES: Conversion from OpenSL ES positional channel mask 0x4 to Android mask 0 loses channels
05-18 16:05:05.013 6572-6572/com.google.sample.echo I/libOpenSLES: Emulating old channel mask behavior (ignoring positional mask 0x4, using default mask 0x10 based on channel count of 1)
05-18 16:05:05.026 6572-6635/com.google.sample.echo E/SchedPolicy: sched_setscheduler failed: tid 6635, errno=1

**Comment by IgorGanapolsky** [comment]
Ok, I was able to get the audio to play on the Galaxy S4 by changing **AUDIO_SAMPLE_CHANNELS               1** to **10**.  It is very choppy though.  At least it plays sound every time.

**Comment by ggfan** [comment]
Adding a fix in https://github.com/googlesamples/android-ndk/pull/410 for echo sample, works on
  nexus 4 - JB /Nexus 5X - N/ Pixel XL - Android O.
When the buffer is too big, it would mean that there is no low latency path --  the size needs to be DMA size between SLES <--> audio driver; I did not see this though on Nexus phones.
@lakdong  if you could try it out on your Nexus 6P, that would be very helpful

Not looking at the Native-Audio problem yet.

**Comment by lakdong** [comment]
echo sample works on Nexus 6p !!

**Comment by ggfan** [comment]
thanks, will get the change reviewed!

**Comment by ggfan** [comment]
merged into main.  thank you for reporting the issue:  will find some time to improve the sample concerning
    UI
    Synchronization etc.

**Comment by ggfan** [comment]
as for Galaxy S4 phone, the reported buffer size is 960 on android L.  The audio path from vendor's OS build may not use the AOSP's source as is: they might have their own audio path.

In the sample, the playback begins after captures 3 buffers, 3 x 960 / 48000Hz = 60ms, then plus platform delays ( mic --> audio driver --> SLES --> audio driver --> speakers: analog ), the latency is felt.  Even I tried to use 2 buffers, do not feel too much difference from3 buffer case, then my guess is that the other part of the system for latency is more significant than from these 2 buffer introduced latency. Closing down this as native-audio fix also merged in.
